File: walkgen/FootStepPlanner.py
#!/usr/bin/env python3
#
# Copyright 2022 University of Edinburgh
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#  * Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of  nor the names of its contributors may be used to
#    endorse or promote products derived from this software without specific
#    prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import numpy as np
import pinocchio as pin
import copy

from walkgen.tools.geometry_utils import Surface
from walkgen.tools.optimisation import quadprog_solve_qp


class FootStepPlanner():
    """ FootStepPlanner initialized by URDF model of the environment.
    Use RBPRM as a guide path since the env is available.
    """

    def __init__(self, model, q, debug = False):
        """ Initialize the FootStepPlanner.

        Args:
            - model (pin.model): Pinocchio robot model.
            - q (array x19): Initial state of the robot.
            - heightmap_path (str): Heightmap binary file path.
            - debug (bool): Store the footstep computed for debug purpose.
        """
        if q.shape[0] != 19:
            raise ArithmeticError(
                "Current state q should be an array of size 19 [pos (x3), quaternion (x4), joint (x12)]")
        # Create the model and data for forward simulation
        self._model = model
        self._data = self._model.createData()

        pin.forwardKinematics(self._model, self._data, q)
        pin.updateFramePlacements(self._model, self._data)

        lf = "LF_FOOT"
        lh = "LH_FOOT"
        rf = "RF_FOOT"
        rh = "RH_FOOT"
        self._contactNames = [lf, lh, rf, rh]
        self._contact_names_SL1M = [lf, rf, lh, rh]
        self._offsets_feet = np.zeros((3, 4))  # Reference shoulder position, base frame.
        self._current_position = np.zeros((3, 4))  # Feet positions, world frame.
        self._current_velocities = np.zeros((3, 4))  # Feet velocities, world frame.
        for i, name in enumerate(self._contactNames):
            Id = self._model.getFrameId(name)
            self._offsets_feet[:, i] = self._data.oMf[Id].translation
            self._current_position[:, i] = self._data.oMf[Id].translation

        self._k_feedback = 0.03
        self._href = 0.48
        self._g = 9.81
        self._L = 0.5

        self.debug = debug
        if debug:
            self.footstep = [] # List of all computed ftesp in the queue of contact for debug purpose

    def compute_footstep(self, queue_cs, q, vq, bvref, timeline, selected_surfaces):
        """ Run the queue in reverse order and update the position for each Contact Schedule (CS)
        following the Raibert heuristic, depending on current linear velocity and both desired angular and
        linear velocty.

        Args :
            - queue_cs (list): List of CS.
            - q (array x19): Current state of the robot.
            - vq (array x18): Linear and angular current velocity.
            - bvref (array x6): Linear and angular desired velocities.
            - timeline (int): timeline in CS.
            - selected_surfaces (dict): Dictionary containing the incoming footstep surfaces.

        Returns:
            - (array 3x4): Target for the incoming footseps.
        """
        if q.shape[0] != 19:
            raise ArithmeticError(
                "Current state q should be an array of size 19 [pos (x3), quaternion (x4), joint (x12)]")
        if vq.shape[0] != 18:
            raise ArithmeticError(
                "Current velocity vq should be an array of size 18 [lin vel (x3), ang vel (x3), joint vel(x12)]")
        if bvref.shape[0] != 6:
            raise ArithmeticError("Reference velocity should be an array of size 6 [lin vel (x3), ang vel (x3)]")

        # Update current feet position
        self._update_current_state(q, vq)

        # Update position for each CS in the queue.
        return self.update_position(queue_cs, q, vq, bvref, timeline, selected_surfaces)

    def update_position(self, queue_cs, q, bv, bvref, timeline_, selected_surfaces):
        """ Update the position for a contact schedule.

        Args:
            - queue_cs (list): List of CS.
            - q (array x19): Current state of the robot.
            - vq (array x18): Linear and angular current velocity.
            - bvref (array x6): Linear and angular desired velocities.
            - timeline (int): timeline in CS.
            - selected_surfaces (dict): Dictionary containing the incoming footstep surfaces.

        Returns:
            - (array 3x4): Target for the incoming footseps.
        """
        if self.debug:
            self.footstep.clear()
            self.footstep = [[],[],[],[]]
            for j in range(4):
                self.footstep[j].append(copy.deepcopy(self._current_position[:,j].tolist()))
        # Get current orientation of the robot
        rpy = pin.rpy.matrixToRpy(pin.Quaternion(q[3:7]).toRotationMatrix())
        Rz = pin.rpy.rpyToMatrix(np.array([0., 0., rpy[2]]))  # Yaw rotation matrix
        Rxy = pin.rpy.rpyToMatrix(np.array([rpy[0], rpy[1], 0.]))  # Roll/Pitch rotation matrix

        q_tmp = q[:3]  # Tmp state, z = 0.
        q_tmp[2] = 0.
        P0 = copy.deepcopy(self._current_position)
        V0 = self._current_velocities.copy()
        timeline = timeline_
        counter = 0

        foot_timeline = [0, 0, 0, 0]
        target_footstep = np.zeros((3,4))
        for cs in reversed(queue_cs):

            for c in range(cs.C):
                name = cs.contactNames[c]
                j = self._contactNames.index(name)
                phases = cs.phases[c]
                N_phase = len(phases)
                if N_phase == 3:
                    T_stance = (phases[0].T + phases[2].T) * cs.dt
                    active_phase = phases[0]
                    inactive_phase = phases[1]

                    # Displacement following the reference velocity compared to current position
                    if active_phase.T + inactive_phase.T - timeline > 0:  # case 1 & 2
                        dt_ = (active_phase.T + inactive_phase.T - timeline) * cs.dt
                        if bvref[5] > 10e-5:
                            dx_ = (bvref[0] * np.sin(bvref[5] * dt_) + bvref[1] *
                                   (np.cos(bvref[5] * dt_) - 1.0)) / bvref[5]
                            dy_ = (bvref[1] * np.sin(bvref[5] * dt_) - bvref[0] *
                                   (np.cos(bvref[5] * dt_) - 1.0)) / bvref[5]
                            yaw = bvref[5] * dt_
                            Rz_tmp = pin.rpy.rpyToMatrix(np.array([0., 0., yaw]))
                        else:
                            dx_ = bvref[0] * dt_
                            dy_ = bvref[1] * dt_
                            Rz_tmp = np.identity(3)

                        q_dxdy = np.array([dx_, dy_, 0.])
                        heuristic = self.compute_heuristic(bv, bvref, Rxy, T_stance, name,
                                                           feedback_term=False)  # without feedback term
                        footstep = Rz @ Rz_tmp @ heuristic + q_tmp + Rz @ q_dxdy

                        P_ = np.identity(3)
                        q_ = np.zeros(3)
                        sf = selected_surfaces.get(name)[foot_timeline[j]]
                        G_ = sf.A
                        h_ = sf.b - sf.A @ footstep
                        delta_x = quadprog_solve_qp(P_, q_, G_, h_)
                        footstep_optim = footstep + delta_x
                        if foot_timeline[j] == 0:
                            target_footstep[:,self._contact_names_SL1M.index(name)] = footstep_optim

                        foot_timeline[j] += 1

                        # else: # if heightmap
                        #     footstep_optim = footstep
                        #     footstep_optim[2] = self.heightmap.get_height(footstep[0], footstep[1])

                        # With feedback term
                        # heuristic_fb = self.compute_heuristic(bv, bvref, Rxy, T_stance, name , feedback_term = True) # with feedback term
                        # footstep_fb = Rz @ Rz_tmp @ heuristic + q_tmp + Rz @ q_dxdy
                        # optimVector.append(OptimData(0,name, selected_surfaces.get(name),heuristic, Rz_tmp ))

                        # Update the trajectory
                        if active_phase.T - timeline >= 0:
                            t0 = 0.
                            V0[:, j] = np.zeros(3)
                        else:
                            t0 = timeline - active_phase.T

                        if t0 <= inactive_phase.T /2:
                            phases[1].trajectory.update(P0[:, j], V0[:,j], footstep_optim, t0 * cs.dt, True)

                        P0[:, j] = footstep_optim
                        V0[:, j] = np.zeros(3)

                        if self.debug:
                            self.footstep[j].append(footstep_optim.tolist())

                    else:  # case 3
                        V0[:, j] = np.zeros(3)
                        if self.debug:
                            self.footstep[j].append(None)

                else:
                    raise ArithmeticError("Problem nombre de phases inside CS.")

            timeline = 0.
            counter += 1

        return target_footstep

    def compute_heuristic(self, bv, bvref, Rxy, T_stance, name, feedback_term=True):
        """ Compute heuristic position in base frame
        """
        footstep = np.zeros(3)

        # Add symmetry term
        footstep += T_stance * 0.5 * bvref[:3]

        # Add feedback term
        footstep += self._k_feedback * bv[:3]
        if feedback_term:
            footstep += -self._k_feedback * bvref[:3]

        #  Add centrifugal term
        cross = np.array([bv[1] * bvref[5] - bv[2] * bvref[4], bv[2] * bvref[3] - bv[0] * bvref[5], 0.0])
        footstep += 0.5 * np.sqrt(self._href / self._g) * cross

        # Legs have a limited length so the deviation has to be limited
        footstep[0] = min(footstep[0], self._L)
        footstep[0] = max(footstep[0], -self._L)
        footstep[1] = min(footstep[1], self._L)
        footstep[1] = max(footstep[1], -self._L)

        # Add shoulders, Yaw axis taken into account later
        j = self._contactNames.index(name)
        footstep += Rxy @ self._offsets_feet[:, j]

        # Remove Z component (working on flat ground)
        footstep[2] = 0.

        return footstep

    def _update_current_state(self, q, vq):
        """ Get current feet positions and velocities by forward kinematics.

        Args:
            - q (array x19): Current state of the robot.
            - bv (array x18): Linear and angular current velocity.
        """
        pin.forwardKinematics(self._model, self._data, q, vq)
        for i, name in enumerate(self._contactNames):
            frame_id = self._model.getFrameId(name)
            oMf = pin.updateFramePlacement(self._model, self._data, frame_id)
            v = pin.getFrameVelocity(self._model, self._data, frame_id)
            self._current_position[:, i] = oMf.translation[:]
            self._current_velocities[:, i] = v.linear[:]

        return 0


if __name__ == "__main__":
    """ Run a simple example of Footstepplanner.
    """
    import os
    from example_robot_data.robots_loader import ANYmalLoader
    from walkgen.GaitManager import GaitManager

    # Reference velocity
    bvref = np.zeros(6)
    bvref[5] = 0.06
    bvref[0] = 0.06

    # Timeline
    timeline = 10

    # Heightmap path
    heightmap_path = os.getcwd() + "/data/lab_scene.dat"

    # Load Anymal model to get the current feet position by forward kinematic.
    ANYmalLoader.free_flyer = True
    anymal = ANYmalLoader().robot

    # FootStepPlanner
    foostep_planner = FootStepPlanner(anymal.model, anymal.q0, heightmap_path)

    cs1, cs2 = dict(), dict()
    cs1["LH_FOOT"] = pin.SE3(np.eye(3), np.array([-0.34, 0.33, 0.0083]))
    cs1["LF_FOOT"] = pin.SE3(np.eye(3), np.array([0.34, 0.23, 0.0083]))
    cs1["RH_FOOT"] = pin.SE3(np.eye(3), np.array([-0.34, -0.23, 0.0083]))
    cs1["RF_FOOT"] = pin.SE3(np.eye(3), np.array([0.34, -0.23, 0.0083]))
    cs2["LH_FOOT"] = pin.SE3(np.eye(3), np.array([0.07, 0.2, 0.0083]))
    cs2["LF_FOOT"] = pin.SE3(np.eye(3), np.array([0.67, 0.2, 0.0083]))
    cs2["RH_FOOT"] = pin.SE3(np.eye(3), np.array([0.07, -0.45, 0.0083]))
    cs2["RF_FOOT"] = pin.SE3(np.eye(3), np.array([0.67, -0.2, 0.0083]))

    gait_manager = GaitManager(anymal.model, anymal.q0)

    # Initial selected surface
    A = [[-1.0000000, 0.0000000, 0.0000000], [0.0000000, -1.0000000, 0.0000000], [0.0000000, 1.0000000, 0.0000000],
         [1.0000000, 0.0000000, 0.0000000], [0.0000000, 0.0000000, 1.0000000], [-0.0000000, -0.0000000, -1.0000000]]

    b = [1.3946447, 0.9646447, 0.9646447, 0.2, 0.1, -0.1]

    vertices = [[-1.3946447276978748, 0.9646446609406726, 0.0], [-1.3946447276978748, -0.9646446609406726, 0.0],
                [0.5346445941834704, -0.9646446609406726, 0.0], [0.5346445941834704, 0.9646446609406726, 0.0]]

    sf = Surface(np.array(A), np.array(b), np.array(vertices).T)

    selected_surfaces = dict()  # Mimic asynchronous behaviour
    for foot in range(4):
        selected_surfaces[foostep_planner._contactNames[foot]] = [sf, sf, sf] # horizon lenght of the Surface is 3.
    foostep_planner.compute_footstep(gait_manager.get_cs(), anymal.q0, anymal.v0, bvref, timeline, selected_surfaces)

    # TODO Add vizualisation.

File: walkgen_footstep_planner/src/FootStepPlanner.cpp
#include "FootStepPlanner.hpp"

#include "pinocchio/math/rpy.hpp"
#include "pinocchio/spatial/se3.hpp"
#include <pinocchio/algorithm/frames.hpp>
#include <pinocchio/math/quaternion.hpp>
#include <pinocchio/multibody/data.hpp>
#include <pinocchio/multibody/model.hpp>

FootStepPlanner::FootStepPlanner(const pinocchio::Model &model, const Eigen::VectorXd &q, const Params &params,
                                 const double period)
    : params_(params), model_(model), period_(period) {
  data_ = pinocchio::Data(model_);
  pinocchio::forwardKinematics(model_, data_, q);
  pinocchio::updateFramePlacements(model_, data_);
  filter_q = FilterMean(period_, static_cast<double>(params_.dt) * params_.nsteps);
  filter_v = FilterMean(period_, static_cast<double>(params_.dt) * params_.nsteps);
  initialize(q);
}

FootStepPlanner::FootStepPlanner(const pinocchio::Model &model, const Eigen::VectorXd &q, const double period)
    : params_(Params()), model_(model), period_(period) {
  data_ = pinocchio::Data(model_);
  pinocchio::forwardKinematics(model_, data_, q);
  pinocchio::updateFramePlacements(model_, data_);
  filter_q = FilterMean(period_, static_cast<double>(params_.dt) * params_.nsteps);
  filter_v = FilterMean(period_, static_cast<double>(params_.dt) * params_.nsteps);
  initialize(q);
}

void FootStepPlanner::initialize(const Eigen::VectorXd &q) {
  // const std::string lf = params_.feet_names[0];
  // const std::string lh = params_.feet_names[2];
  // const std::string rf = params_.feet_names[1];
  // const std::string rh = params_.feet_names[3];

  contactNames_ = params_.feet_names;
  contactNames_sl1m_ = params_.feet_names_sl1m;

  // TODO find a better way for this
  // offsets_feet_[contactNames_[0]] = Vector3(0.367, 0.2, 0.);
  // offsets_feet_[contactNames_[1]] = Vector3(0.367, -0.2, 0.);
  // offsets_feet_[contactNames_[2]] = Vector3(-0.367, 0.2, 0.);
  // offsets_feet_[contactNames_[3]] = Vector3(-0.367, -0.2, 0.);

  if (contactNames_.size() != params_.shoulder_offsets.size()) {
    throw std::runtime_error(
        "The contact name list and the shoulder offsets should "
        "have the same size.");
  }
  if (contactNames_sl1m_.size() != contactNames_.size()) {
    throw std::runtime_error(
        "The size of each offset in shoulder offset params "
        "should be size 2 (x and y axis).");
  }
  for (auto elem : params_.shoulder_offsets) {
    if (elem.size() != 2) {
      throw std::runtime_error(
          "The size of each offset in shoulder offset params "
          "should be size 2 (x and y axis).");
    }
  }
  for (size_t j = 0; j < contactNames_.size(); j++) {
    Vector3 offset = Vector3::Zero();
    offset[0] = params_.shoulder_offsets[j][0];
    offset[1] = params_.shoulder_offsets[j][1];
    offsets_feet_[contactNames_[j]] = offset;
  }
  std::sort(contactNames_.begin(), contactNames_.end());

  current_position_ = MatrixN::Zero(3, 4);
  current_velocities_ = Matrix34::Zero();
  target_fsteps_ = Matrix34::Zero();
  qf_ = VectorN::Zero(18);
  qvf_ = VectorN::Zero(18);

  nsteps_ = params_.nsteps;
  horizon_ = params_.horizon;
  dt_ = params_.dt;
  counter_gait_ = 0;
  early_termination_ratio_ = params_.early_termination_ratio;

  double dx = 0.5;
  double dy = 0.5;
  double height = 0.;
  double epsilon = 10e-6;
  Eigen::Matrix<double, 6, 3> A;
  A << -1., 0., 0., 0., -1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 1., 0., 0., -1.;
  Eigen::Matrix<double, 6, 1> b;
  b << dx - q(0), dy - q(1), dy + q(1), dx + q(0), height + epsilon, -height + epsilon;

  Eigen::Matrix<double, 4, 3> vertices;
  vertices << q[0] - dx, q[1] + dy, height, q[0] - dx, q[1] - dy, height, q[0] + dx, q[1] - dy, height, q[0] + dx,
      q[1] + dy, height;
  for (size_t foot = 0; foot < 4; ++foot) {
    previous_surfaces_[contactNames_[foot]] = Surface(MatrixN(A), b, MatrixN(vertices));
  }

  // Initialize the tmp variables
  oMf_tmp = pinocchio::SE3::Identity();
  v_tmp = pinocchio::Motion::Zero();
  footstep_tmp.setZero();
  cross_tmp.setZero();
  q_filter_tmp.setZero();
  Rz.setZero();
  Rxy.setZero();
  q_tmp.setZero();
  P0.setZero();
  V0.setZero();
  cs_index = 0;
  timeline = 0;
  foot_timeline = {0, 0, 0, 0};
  target_fsteps = MatrixN::Zero(3, 4);
  Rz_tmp = Matrix3::Identity();
  dt_i = 0.;
  dx = 0.;
  dy = 0.;
  q_dxdy.setZero();
  C_ = MatrixN::Zero(3, 0);
  d_ = VectorN::Zero(0);
  delta_x = VectorN::Zero(3);
  G_ = MatrixN::Zero(3, 3);  // Random size
  h_ = VectorN::Zero(3);
  fsteps_optim.setZero();
  previous_sf = Surface(MatrixN(A), b, MatrixN(vertices));
};

// From python, it is easier to use and return only MatrixN, otherwise, a
// binding is necessary for each type.
MatrixN FootStepPlanner::compute_footstep(std::vector<std::shared_ptr<ContactSchedule>> &queue_cs, const VectorN &q,
                                          const VectorN &vq, const VectorN &bvref, const int timeline,
                                          const std::map<std::string, std::vector<Surface>> &selected_surfaces,
                                          const std::map<std::string, Surface> &previous_surfaces) {
  if (q.size() != 19) {
    throw std::runtime_error(
        "Current state q should be an array of size 19 "
        "[pos (x3), quaternion (x4), joint (x12)]");
  }
  if (vq.size() != 18) {
    throw std::runtime_error(
        "Current velocity vq should be an array of size "
        "18 [lin vel (x3), ang vel (x3), joint vel(x12)]");
  }
  if (bvref.size() != 6) {
    throw std::runtime_error(
        "Reference velocity should be an array of size 6 "
        "[lin vel (x3), ang vel (x3)]");
  }

  // Update current feet position
  update_current_state(q, vq);

  // Filter quantities
  q_filter_tmp.head<3>() = q.head<3>();
  q_filter_tmp.tail<3>() =
      pinocchio::rpy::matrixToRpy(pinocchio::SE3::Quaternion(q(6), q(3), q(4), q(5)).toRotationMatrix());

  qf_ = filter_q.filter(q_filter_tmp);
  qvf_ = filter_v.filter(vq.head<6>());

  return update_position(queue_cs, qf_, qvf_, bvref, timeline, selected_surfaces);
}

MatrixN FootStepPlanner::update_position(std::vector<std::shared_ptr<ContactSchedule>> &queue_cs, const Vector6 &q,
                                         const Vector6 &vq, const VectorN &bvref, const int timeline_in,
                                         const std::map<std::string, std::vector<Surface>> &selected_surfaces) {
  if (timeline_in == 0) {
    counter_gait_ += 1;
  }
  Rz = pinocchio::rpy::rpyToMatrix(double(0.), double(0.), q(5));
  Rxy = pinocchio::rpy::rpyToMatrix(q(3), q(4), double(0.));

  q_tmp = q.head<3>();
  q_tmp(2) = double(0.);

  P0 = current_position_;
  V0 = current_velocities_;
  timeline = timeline_in;

  // Reset quantities
  cs_index = 0;
  std::fill(foot_timeline.begin(), foot_timeline.end(), 0);
  target_fsteps.setZero();
  Rz_tmp.setIdentity();

  for (auto cs_iter = queue_cs.rbegin(); cs_iter != queue_cs.rend(); ++cs_iter) {
    auto &cs = **cs_iter;  // Reference to ContactSchedule
    if (cs_index <= horizon_ + 2) {
      for (size_t c = 0; c < cs.contactNames_.size(); ++c) {
        auto &name = cs.contactNames_[c];
        size_t j = find_stdVec(cs.contactNames_, name);  // Which foot in foot_timeline
        auto &phases = cs.phases_[c];
        if (phases.size() == 3) {
          double T_stance = double(phases[0]->T_ + phases[2]->T_) * cs.dt_;
          auto &active_phase = phases[0];
          auto &inactive_phase = phases[1];

          if (cs_index + active_phase->T_ - timeline <= horizon_ + 2) {
            // Displacement following the reference velocity compared to current
            // position
            if (active_phase->T_ + inactive_phase->T_ - timeline > 0) {  // case 1 and 2
              if (std::abs(bvref(5)) > 0.01) {
                dt_i = static_cast<double>(cs_index + active_phase->T_ + inactive_phase->T_ - timeline) *
                       cs.dt_;  // dt integration dt_i
                dx = (bvref(0) * std::sin(bvref(5) * dt_i) + bvref(1) * (std::cos(bvref(5) * dt_i) - 1.0)) / bvref(5);
                dy = (bvref(1) * std::sin(bvref(5) * dt_i) - bvref(0) * (std::cos(bvref(5) * dt_i) - 1.0)) / bvref(5);
                dt_i = static_cast<double>(cs_index + active_phase->T_ + inactive_phase->T_) * cs.dt_;
                Rz_tmp = pinocchio::rpy::rpyToMatrix(double(0.), double(0.), bvref(5) * dt_i);
              } else {
                dt_i = static_cast<double>(cs_index + active_phase->T_ + inactive_phase->T_ - timeline) * cs.dt_;
                dx = bvref(0) * dt_i;
                dy = bvref(1) * dt_i;
                Rz_tmp = Matrix3::Identity();
              }
              q_dxdy << dx, dy, double(0.);
              heuristic_tmp = compute_heuristic(vq, bvref, Rxy, T_stance, name, false);
              footstep_ref = Rz * (Rz_tmp * heuristic_tmp) + q_tmp + Rz * q_dxdy;

              Matrix3 P_ = Matrix3::Identity();
              Vector3 q_ = Vector3::Zero();
              auto iter = selected_surfaces.find(name);
              if (iter == selected_surfaces.end()) {
                throw std::runtime_error("Naming non consistent in surface dictionnary.");
              }
              auto &sf_ = (iter->second).at(foot_timeline.at(j));
              G_ = -sf_.getA();                             // Inverse sign in python
              h_ = sf_.getb() - sf_.getA() * footstep_ref;  // Inverse sign in python

              delta_x.setZero();
              // In python code
              // delta_x = quadprog_solve_qp(P_, q_, G_, h_)
              // min (1/2)x' P x + q' x
              // subject to  G x <= h
              // subject to  C x  = d

              // Eiquadprog-Fast solves the problem :
              // min. 1/2 * x' P_ x + q_' x
              // s.t. C_ x + d_ = 0
              //      G_ x + h_ >= 0
              status = qp.solve_quadprog(P_, q_, C_, d_, G_, h_, delta_x);
              fsteps_optim = footstep_ref + delta_x;

              // Update target fsteps for sl1m
              if (foot_timeline.at(j) == 0) {
                size_t jj = find_stdVec(contactNames_sl1m_, name);
                target_fsteps_.col(static_cast<Eigen::Index>(jj)) = fsteps_optim;
              }

              // Update previous surface mechanism
              if (foot_timeline.at(j) == 0) {
                auto iter = previous_surfaces_.find(name);
                if (iter == previous_surfaces_.end()) {
                  throw std::runtime_error("Naming non consistent in previous_surface dictionnary.");
                }
                previous_sf = Surface(iter->second);  // make a copy
              } else {
                auto iter = selected_surfaces.find(name);
                if (iter == selected_surfaces.end()) {
                  throw std::runtime_error("Naming non consistent in selected_surface dictionnary.");
                }
                previous_sf = Surface((iter->second).at(foot_timeline.at(j) - 1));
              }
              double t0 = double(0.);

              // Update trajectory
              if (active_phase->T_ - timeline >= 0) {
                V0.col(static_cast<Eigen::Index>(j)).setZero();
              } else {
                t0 = static_cast<double>(timeline - active_phase->T_);
              }
              if (counter_gait_ < 2) {
                // Letting 1/2 gait for the estimator at first since it is a
                // moving average estimator. Hence, keeping initial position.
                fsteps_optim.head<2>() = P0.col(static_cast<Eigen::Index>(j)).head<2>();
              }
              if (t0 <= static_cast<double>(inactive_phase->T_) * early_termination_ratio_) {
                inactive_phase->trajectory_->update(P0.col(static_cast<Eigen::Index>(j)),
                                                    V0.col(static_cast<Eigen::Index>(j)), fsteps_optim, t0 * cs.dt_,
                                                    sf_, previous_sf);
              }

              // End of the flying phase, register the surface.
              if (t0 >= static_cast<double>(inactive_phase->T_ - nsteps_)) {
                // previous_surfaces_ .pop(name)
                auto iter = previous_surfaces_.find(name);
                if (iter == previous_surfaces_.end()) {
                  throw std::runtime_error("Naming non consistent in previous_surface dictionnary.");
                }
                iter->second = Surface(sf_);  // make a copy.
              }
              P0.col(static_cast<Eigen::Index>(j)) = fsteps_optim;
              V0.col(static_cast<Eigen::Index>(j)).setZero();

              foot_timeline.at(j) += 1;
            } else {  // case 3
              V0.col(static_cast<Eigen::Index>(j)).setZero();
            }
          }
        } else {
          throw std::runtime_error(
              "Error in fstep planner. Only 3 phases per "
              "ContactSchedule considered.");
        }
      }
    } else {
      break;
    }
    cs_index += cs.T_ - timeline;
    timeline = 0;
  }

  return target_fsteps_;
}

void FootStepPlanner::update_current_state(const Eigen::VectorXd &q, const Eigen::VectorXd &vq) {
  if (q.size() != 19) {
    throw std::runtime_error(
        "Current state q should be an array of size 19 "
        "[pos (x3), quaternion (x4), joint (x12)]");
  }
  if (vq.size() != 18) {
    throw std::runtime_error(
        "Current velocity vq should be an array of size "
        "18 [lin vel (x3), ang vel (x3), joint vel(x12)]");
  }
  pinocchio::forwardKinematics(model_, data_, q, vq);
  size_t frame_id;
  for (size_t i = 0; i < contactNames_.size(); i++) {
    frame_id = model_.getFrameId(contactNames_[i]);
    oMf_tmp = pinocchio::updateFramePlacement(model_, data_, frame_id);
    v_tmp = pinocchio::getFrameVelocity(model_, data_, frame_id);
    current_position_.col(static_cast<Eigen::Index>(i)) = oMf_tmp.translation();
    current_velocities_.col(static_cast<Eigen::Index>(i)) = v_tmp.linear();
  }
}

Vector3 FootStepPlanner::compute_heuristic(const Eigen::VectorXd &bv, const Eigen::VectorXd &bvref, const Matrix3 &Rxy,
                                           const double T_stance, const std::string &name, const bool feedback_term) {
  // Reset tmp variables.
  footstep_tmp.setZero();
  cross_tmp.setZero();

  // Add symmetry term
  footstep_tmp += beta_ * T_stance * bvref.head<3>();

  // Add feedback term
  if (feedback_term) {
    footstep_tmp += k_feedback_ * bv.head<3>();
    footstep_tmp += -k_feedback_ * bvref.head<3>();
  }

  // Add centrifugal term
  Vector3 cross;
  cross_tmp << bv(1) * bvref(5) - bv(2) * bvref(4), bv(2) * bvref(3) - bv(0) * bvref(5), 0.0;
  footstep_tmp += 0.5 * std::sqrt(href_ / g_) * cross_tmp;

  // Limit deviation
  footstep_tmp(0) = std::min(footstep_tmp(0), L_);
  footstep_tmp(0) = std::max(footstep_tmp(0), -L_);
  footstep_tmp(1) = std::min(footstep_tmp(1), L_);
  footstep_tmp(1) = std::max(footstep_tmp(1), -L_);

  // Add shoulders, Yaw axis taken into account later
  // size_t j = find_stdVec(contactNames_,name);
  footstep_tmp += Rxy * offsets_feet_.at(name);

  // Remove Z component (working on flat ground)
  footstep_tmp(2) = 0.0;

  return footstep_tmp;
}

// find element inside vector list.
template <typename Type>
size_t FootStepPlanner::find_stdVec(const std::vector<Type> &vec, const Type &elem) {
  auto it = std::find(vec.begin(), vec.end(), elem);
  if (it != vec.end()) {
    return static_cast<size_t>(std::distance(vec.begin(), it));
  } else {
    throw std::runtime_error("Element not found in vector");
  }
}
